test_scripts/Security/GetSystemTime/common.lua

```diff
@@ -0,0 +1,164 @@
+    tz_hour = 2,
+    tz_minute = 0
+  }
+end
+
+function m.setForceProtectedServiceParam(pParamValue)
+  m.setSDLIniParameter("ForceProtectedService", pParamValue)
+end
+
+function m.getAppID(pAppId)
+  return m.getConfigAppParams(pAppId).appID
+end
+
+local function allowSDL()
+  test.hmiConnection:SendNotification("SDL.OnAllowSDLFunctionality", {
+    allowed = true,
+    source = "GUI",
+    device = {
+      id = utils.getDeviceMAC(),
+      name = utils.getDeviceName()
+    }
+  })
+end
+
+function m.start(pOnSystemTime, pHMIParams)
+  test:runSDL()
+  commonFunctions:waitForSDLStart(test)
+  :Do(function()
+      test:initHMI()
+      :Do(function()
+          utils.cprint(35, "HMI initialized")
+          test:initHMI_onReady(pHMIParams)
+          :Do(function()
+              utils.cprint(35, "HMI is ready")
+              if pOnSystemTime then
+                m.getHMIConnection():SendNotification("BasicCommunication.OnSystemTimeReady")
+              end
+              test:connectMobile()
+              :Do(function()
+                  utils.cprint(35, "Mobile connected")
+                  allowSDL(test)
+                end)
+            end)
+        end)
+    end)
+end
+
+function m.expectHandshakeMessage(pGetSystemTimeOccur, pTime, pHandshakeOccurences)
+  if not pTime then
+    pTime = getSystemTimeValue()
+  end
+  if not pHandshakeOccurences then pHandshakeOccurences = 1 end
+  if pGetSystemTimeOccur == 0 then
+    pHandshakeOccurences = pGetSystemTimeOccur
+  end
+  m.getMobileSession():ExpectHandshakeMessage()
+    :Times(pHandshakeOccurences)
+  EXPECT_HMICALL("BasicCommunication.GetSystemTime")
+  :Do(function(_, d)
+    m.getHMIConnection():SendResponse(d.id, d.method, "SUCCESS", { systemTime = pTime })
+  end)
+  :Times(pGetSystemTimeOccur)
+end
+
+function m.startServiceSecured(pData, pServiceId, pGetSystemTimeOccur, pTime)
+  m.getMobileSession():StartSecureService(pServiceId)
+  m.getMobileSession():ExpectControlMessage(pServiceId, pData)
+
+  m.getHMIConnection():ExpectNotification("SDL.OnStatusUpdate")
+  :Times(0)
+
+  m.expectHandshakeMessage(pGetSystemTimeOccur, pTime)
+end
+
+local function expNotDuringPTU()
+  m.getHMIConnection():ExpectRequest("VehicleInfo.GetVehicleData", { odometer = true })
+end
+
+function m.startServiceSecuredwithPTU(pData, pServiceId, pGetSystemTimeOccur, pTime, pPTUpdateFunc, pHandshakeOccurences)
+  m.getMobileSession():StartSecureService(pServiceId)
+  m.getMobileSession():ExpectControlMessage(pServiceId, pData)
+
+  m.getHMIConnection():ExpectNotification("SDL.OnStatusUpdate",
+    { status = "UPDATE_NEEDED" }, { status = "UPDATING" }, { status = "UP_TO_DATE" })
+    :Times(3)
+    :Do(function(e)
+      if e.occurences == 1 then
+        m.policyTableUpdate(pPTUpdateFunc, expNotDuringPTU)
+      end
+    end)
+
+  m.expectHandshakeMessage(pGetSystemTimeOccur, pTime, pHandshakeOccurences)
+end
+
+function m.startServiceSecuredWitTimeoutWithoutGetSTResp(pData, pServiceId, pTimeout)
```

Comment:
On further inspection, it seems to me like this should be split into two functions, one with a delayed response for `GetSystemTime` and one where no response is sent for `GetSystemTime`. Combining the two seems more confusing than anything, since the presence of the timeout parameter changes the flow quite significantly (although you could use local functions to combine these is you want).

Reply:
@jacobkeeler Please check 3758adb